torch/_refs/linalg/__init__.py

```diff
@@ -0,0 +1,129 @@
+import torch
+from torch import Tensor
+
+import torch._prims as prims
+import torch._refs as refs
+from torch._prims.wrappers import out_wrapper
+
+from torch._prims.utils import (
+    check,
+    DimsType,
+    TensorLikeType,
+    NumberType,
+    corresponding_real_dtype,
+    is_float_dtype,
+    is_complex_dtype,
+    get_computation_dtype,
+    get_higher_dtype,
+    reduction_dtypes,
+    REDUCTION_OUTPUT_TYPE_KIND,
+)
+import torch._refs.linalg as linalg
+import torch._refs.linalg.utils
+
+from typing import Optional, List
+from functools import partial
+
+__all__ = [
+    "vector_norm",
+]
+
+
+def check_norm_dtype(dtype: Optional[torch.dtype], x_dtype: torch.dtype, fn_name: str):
+    """
+    Checks related to the dtype kwarg in `linalg.*norm` functions
+    """
+    if dtype is not None:
+        check(
+            is_float_dtype(dtype) or is_complex_dtype(dtype),
+            lambda: f"{fn_name}: dtype should be floating point or complex. Got {dtype}",
+        )
+        check(
+            is_complex_dtype(dtype) == is_complex_dtype(x_dtype),
+            lambda: "{fn_name}: dtype should be {d} for {d} inputs. Got {dtype}".format(
+                fn_name=fn_name,
+                d="complex" if is_complex_dtype(x_dtype) else "real",
+                dtype=dtype,
+            ),
+        )
+        check(
+            get_higher_dtype(dtype, x_dtype) == dtype,
+            lambda: f"{fn_name}: the dtype of the input ({x_dtype}) should be convertible "
+            "without narrowing to the specified dtype ({dtype})",
+        )
+
+
+@out_wrapper(exact_dtype=True)
+def vector_norm(
+    x: TensorLikeType,
+    ord: float = 2.0,
+    dim: Optional[DimsType] = None,
+    keepdim: bool = False,
+    *,
+    dtype: Optional[torch.dtype] = None,
+) -> Tensor:
+    # Checks
+    linalg.utils.check_fp_or_complex(
+        x.dtype, "linalg.vector_norm", allow_low_precision_dtypes=True
+    )
+
+    if isinstance(dim, int):
+        dim = [dim]  # type: ignore[assignment]
+    elif not isinstance(dim, List) and dim is not None:
+        # refs.sum just accepts List rather than DimType
+        dim = list(dim)  # type: ignore[assignment]
+
+    if x.numel() == 0 and (ord < 0.0 or ord == float("inf")):
+        check(
+            dim is not None and len(dim) != 0,
+            lambda: f"linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
+            "because the operation does not have an identity",
+        )
+        shape = x.shape
+        assert dim is not None  # mypy does not seem to be able to see through check?
+        for d in dim:
```

Comment:
@ngimel asks: Does this work for reducing a 0D tensor over the 0th dimension?

```
a = torch.randn(())
torch.linalg.vector_norm(a, ord=float('inf'), dim=0)
: tensor(0.7286)

a.shape
: torch.Size([])
```

In this case dim is 0 but `shape[d]` does not exist.

We should add a sample input for this case, too.

Reply:
uuuh. This is sharp. I'll add a sample for this and correct it both in the function and the reference. Thanks!

Reply:
Actually nevermind, this works as expected because this if is inside a `numel() == 0`, so it doesn't apply to that example

Note that this is a transcription line by line of the meta function for the operation, so if it works in the operation it should work here.